### PR TITLE
typescript-angular: Feature/model suffix

### PR DESCRIPTION
**Pull request by tomvangreen**
### PR checklist

- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Ran the shell script under `./bin/` to update Petstore sample so that CIs can verify the change. (For instance, only need to run `./bin/{LANG}-petstore.sh` and `./bin/security/{LANG}-petstore.sh` if updating the {LANG} (e.g. php, ruby, python, etc) code generator or {LANG} client's mustache templates). Windows batch files can be found in `.\bin\windows\`.
- [x] Filed the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`, `3.1.x`, `4.0.x`. Default: `master`.
- [x] Copied the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) to review the pull request if your PR is targeting a particular programming language.

### Description of the PR
I introduced a `modelSuffix` and `modelFileSuffix` parameter in order to add suffixes to the generated models and model file names. Taking the samples from the sample, with this setting you could for example change the model `Pet` in the file `pet.ts` to `PetVm` in the file `pet.vm.ts` by defining the model suffix `Vm` and the model file suffix `.vm`.

The motivation behind this change is to have a clear distincation between generated models and models that are created by a developer. Having the ability to add a suffix to the generated models makes it immediatly obvious of what kind a model is.

I currently see one issue and that is that the class suffix gets also appended to enumeration types. The `StatusEnum` in the `Pet` class will become `StatusVmEnum` if I set the `modelSuffix` option to `Vm`. I will have a look if I find how I can avoid that. 

@TiFu @taxpon  @sebastianhaas @kenisteward @Vrolijkx @macjohnny

### CODE REVIEW

**Comment by tomvangreen** [comment]
#234

**Comment by tomvangreen** [comment]
Yes, you are right. You could set the suffix to something invalid. I think this could also be a problem with the filename.

From what I see we have following options:
1. Validate and throw an exception if the name is invalid.
2. Sanitize the input and strip every invalid character and show a warning.
3. Let it run and fail.

I would say because of the filename probably approach 1 or 2 should be used. Better to catch this problem than to let it fail uncontrolled. Do you know of any similar cases in this project and how it is handled there?

**Comment by tomvangreen** [comment]
I added validation for class and file suffixes. As my earlier change with the service suffix has the same problem, I also added the validation for those configurations.

Currently I only allow alpha numeric characters for class suffixes. For file suffixes I additionally allow the characters '.' and '-'. 

If any of the four config values is invalid I throw an IllegalArgument exception and display a message.

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/TypeScriptAngularClientCodegen.java [hunk 1]
```diff
@@ -486,7 +505,44 @@ private String getApiFilenameFromClassname(String classname) {
...
+    }
+
+    private String removeModelSuffixIfNecessary(String name) {
+        if (modelSuffix.length() == 0 || !name.endsWith(modelSuffix)) {
+            return name;
+        }
+        return name.substring(0, name.length() - modelSuffix.length());
     }
 
+    private void validateFileSuffixArgument(String argument, String value) {
```
Comment (macjohnny) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/TypeScriptAngularClientCodegen.java]: please add a method description. describe what `argument` and `value` are or rename them to better understand their meaning

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/TypeScriptAngularClientCodegen.java [hunk 2]
```diff
@@ -486,7 +505,44 @@ private String getApiFilenameFromClassname(String classname) {
...
+        if (!value.matches(FILE_NAME_SUFFIX_PATTERN)) {
+            throw new IllegalArgumentException(
+                    String.format("%s file suffix only allows '.', '-' and alphanumeric characters.", argument)
+            );
+        }
+    }
+
+    private void validateClassSuffixArgument(String argument, String value) {
```
Comment (macjohnny) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/TypeScriptAngularClientCodegen.java]: same here

**Comment by TiFu** [comment]
Agreed. Thanks for the PR!

**Comment by macjohnny** [comment]
@wing328 could you please add this to the `3.1.0` milestone?

**Pull request merged**